### PR TITLE
Harmonize commit return type

### PR DESCRIPTION
**Pull request by Wauplin**
Related to https://github.com/huggingface/datasets/pull/5545#issuecomment-1438557414 (cc @mariosasko @lhoestq).

This PR harmonize the return value between `create_commit`, `upload_file` and `upload_folder`. At the moment, we have a `CommitInfo` dataclass with useful information for `create_commit` (and `delete_file`/`delete_folder`). However, `upload_file` returns a URL to the uploaded file on the Hub while `upload_folder` returns a URL to the uploaded folder on the Hub. This PR makes `CommitInfo` inherits from `str` so that return value can always contain the commit information.

This is somehow similar to the `RepoUrl` when creating a repo. It shouldn't introduce any breaking change to the existing integrations. In particular, `huggingface-cli upload` still prints the same urls as before. In a future version (not before `v1.0`) we should remove the `CommitInfo` x `str` mixin introduced in this PR and settle for a dataclass-only solution.

I have adapted a few tests to check the return value type. I also took the opportunity to simplify the logic in some tests, especially removing some `self.subTest` (we are testing twice the same feature on ~ the same setting) and `self.fail` (which doesn't play nice with `--pdb`) that makes the tests a bit more readable IMO.

### CODE REVIEW

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -4517,27 +4591,39 @@ def upload_folder(
...
+            repo_type=repo_type,
+            repo_id=repo_id,
+            operations=commit_operations,
+            commit_message=commit_message,
+            commit_description=commit_description,
+            token=token,
+            revision=revision,
+            create_pr=create_pr,
+            parent_commit=parent_commit,
+        )
```
Comment (lhoestq) [inline src/huggingface_hub/hf_api.py:4608]: why do you need this extra commit for the multi_commits case ?

Comment (mariosasko) [reply src/huggingface_hub/hf_api.py:4608]: The function returns before this line is the multi-commit case: https://github.com/huggingface/huggingface_hub/blob/b37f20b91af4f5957ef23c4f42d4714a5fb28f40/src/huggingface_hub/hf_api.py#L4581-L4596

File: src/huggingface_hub/hf_api.py [hunk 2]
```diff
@@ -4517,27 +4591,39 @@ def upload_folder(
...
-                commit_description=commit_description,
-                token=token,
-                revision=revision,
-                create_pr=create_pr,
-                parent_commit=parent_commit,
-            )
-            pr_url = commit_info.pr_url
+            # Defining a CommitInfo object is not really relevant in this case
+            # Let's return early with pr_url only (as string).
+            return pr_url
```
Comment (mariosasko) [inline src/huggingface_hub/hf_api.py:4596]: Returning a list of commits in the multi-commit case would make more sense, no?

**Review by mariosasko — COMMENTED** [review]
Nice! One question:

File: src/huggingface_hub/hf_api.py (hunk 2 above)
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:4596]: In general `multi_commits` is still flagged as `"experimental"` and supposedly not much used so I prefer not to touch in this PR to keep it simple. In any case, we can update the type in the future without prior notice. Returning a list of commits is a good idea though. We would need an extra call to list the PR commits but that's straightforward.

File: src/huggingface_hub/hf_api.py (hunk 1 above)
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:4608]: Yes exactly. As @mariosasko said, the multi_commit is processed separately.

**Review by LysandreJik — APPROVED** [review]
So many overloads :face_with_spiral_eyes: 

Great to have everything harmonized.

Very clean! Two literals are manageable but it already starts being a lot! Impressive refactor on the tests.

**Pull request merged**